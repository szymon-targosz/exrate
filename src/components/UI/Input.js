import React from 'react';
import PropTypes from 'prop-types';

const Input = ({ elementType, config, cssClasses, changeHandler, options, label, ...rest }) => {
   let inputElem = null;
   let classes = [ 'input__control' ];

   if (cssClasses) classes = classes.concat(cssClasses);

   switch (elementType) {
      case 'input':
         inputElem = <input className={classes.join(' ')} {...config} onChange={changeHandler} {...rest} />;
         break;

      case 'select':
         inputElem = (
            <select className={classes.join(' ')} {...config} onChange={changeHandler} {...rest}>
               {options.map(option => (
                  <option key={option.value} value={option.value}>
                     {option.text}
                  </option>
               ))}
            </select>
         );
         break;

      default:
         inputElem = <input type="text" className={classes.join(' ')} {...config} onChange={changeHandler} {...rest} />;
         break;
   }

   return (
      <div className="input">
         {inputElem}
         {(label && elementType !== 'select') && <label className="input__label" htmlFor={config.id}>{config.placeholder}</label>}
      </div>
   );
};

Input.propTypes = {
   elementType: PropTypes.string,
   cssClasses: PropTypes.oneOfType([ PropTypes.string, PropTypes.array ]),
   changeHandler: PropTypes.func.isRequired,
   config: PropTypes.shape({
      value: PropTypes.oneOfType([ PropTypes.string, PropTypes.number ]).isRequired
   })
};

export default Input;
