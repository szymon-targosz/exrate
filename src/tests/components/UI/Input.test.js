import React from 'react';
import { shallow } from 'enzyme';
import Input from '../../../components/UI/Input';

let wrapper, changeHandler;

beforeEach(() => {
   changeHandler = jest.fn();
   wrapper = shallow(<Input changeHandler={changeHandler} config={{ value: 'input value' }} />);
});

describe('COMPONENT - Input', () => {
   test('Should render input type "text" if the elementType prop is not defined', () => {
      expect(wrapper).toMatchSnapshot();
   });

   test('Should render input type "text"', () => {
      wrapper.setProps({ elementType: 'input', config: { value: 'input value', type: 'text' } });
      expect(wrapper).toMatchSnapshot();
   });

   test('Should render input type "text" with label', () => {
      wrapper.setProps({
         elementType: 'input',
         config: {
            value: 'input value',
            type: 'text',
            id: 'some-input-id',
            placeholder: 'placeholder text'
         },
         label: true
      });
      expect(wrapper).toMatchSnapshot();
   });

   test('Should render select element', () => {
      wrapper.setProps({
         elementType: 'select',
         options: [ { value: 'asc', text: 'ascending' }, { value: 'desc', text: 'descending' } ]
      });
      expect(wrapper).toMatchSnapshot();
   });

   test('Should add external css classes - array', () => {
      wrapper.setProps({ cssClasses: [ 'bg-secondary', 'border', 'border-info' ] });
      expect(wrapper).toMatchSnapshot();
   });

   test('Should add external css classes - string', () => {
      wrapper.setProps({ cssClasses: 'bg-warning p-s' });
      expect(wrapper).toMatchSnapshot();
   });

   test('Should handle onChange event', () => {
      wrapper.find('input').simulate('change');
      expect(changeHandler).toHaveBeenCalled();
   });
});
