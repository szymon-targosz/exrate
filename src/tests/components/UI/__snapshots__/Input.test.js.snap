// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT - Input Should add external css classes - array 1`] = `
<div
  className="input"
>
  <input
    className="input__control bg-secondary border border-info"
    onChange={[MockFunction]}
    type="text"
    value="input value"
  />
</div>
`;

exports[`COMPONENT - Input Should add external css classes - string 1`] = `
<div
  className="input"
>
  <input
    className="input__control bg-warning p-s"
    onChange={[MockFunction]}
    type="text"
    value="input value"
  />
</div>
`;

exports[`COMPONENT - Input Should render input type "text" 1`] = `
<div
  className="input"
>
  <input
    className="input__control"
    onChange={[MockFunction]}
    type="text"
    value="input value"
  />
</div>
`;

exports[`COMPONENT - Input Should render input type "text" if the elementType prop is not defined 1`] = `
<div
  className="input"
>
  <input
    className="input__control"
    onChange={[MockFunction]}
    type="text"
    value="input value"
  />
</div>
`;

exports[`COMPONENT - Input Should render input type "text" with label 1`] = `
<div
  className="input"
>
  <input
    className="input__control"
    id="some-input-id"
    onChange={[MockFunction]}
    placeholder="placeholder text"
    type="text"
    value="input value"
  />
  <label
    className="input__label"
    htmlFor="some-input-id"
  >
    placeholder text
  </label>
</div>
`;

exports[`COMPONENT - Input Should render select element 1`] = `
<div
  className="input"
>
  <select
    className="input__control"
    onChange={[MockFunction]}
    value="input value"
  >
    <option
      key="asc"
      value="asc"
    >
      ascending
    </option>
    <option
      key="desc"
      value="desc"
    >
      descending
    </option>
  </select>
</div>
`;
